sw/airborne/arch/linux/mcu_periph/pwm_sysfs.c

```diff
@@ -0,0 +1,179 @@
+#else
+#define PS_DEBUG_PRINT(...) {}
+#endif
+
+// Write in file
+static int write_file(const char *path, const char *fmt, ...);
+// Read file content
+//static int read_file(const char *path, const char *fmt, ...);
+
+int pwm_sysfs_init(struct PWM_Sysfs *pwm, char *base_path,
+    char *_export, char *_enable,
+    char *_duty, char *_period,
+    uint8_t channel)
+{
+  // path to export file (create new PWM)
+  char export_path[PWM_SYSFS_PATH_LEN];
+
+  if (base_path == NULL || _export == NULL || _enable == NULL ||
+      _period == NULL || _duty == NULL) {
+    // invalid paths
+    PS_DEBUG_PRINT("invalid paths\n");
+    return -1;
+  }
+  // store paths
+  snprintf(export_path, PWM_SYSFS_PATH_LEN, "%s/%s", base_path, _export);
+  snprintf(pwm->enable_path, PWM_SYSFS_PATH_LEN, "%s/pwm_%u/%s", base_path, channel, _enable);
+  snprintf(pwm->duty_path, PWM_SYSFS_PATH_LEN, "%s/pwm_%u/%s", base_path, channel, _duty);
+  snprintf(pwm->period_path, PWM_SYSFS_PATH_LEN, "%s/pwm_%u/%s", base_path, channel, _period);
+
+  PS_DEBUG_PRINT("%s\n%s\n%s\n%s\n",
+      export_path, pwm->enable_path, pwm->duty_path, pwm->period_path);
+
+  // export new PWM entry
+  write_file(export_path, "%u", channel);
+  // open duty cycle file
+  pwm->duty_cycle_fd = open(pwm->duty_path, O_RDWR | O_CLOEXEC);
+  if (pwm->duty_cycle_fd < 0) {
+    // fail to open duty file
+    PS_DEBUG_PRINT("failed to open FD cycle: %d\n", pwm->duty_cycle_fd);
+    return -2;
+  }
+  // set default period
+  pwm_sysfs_set_period(pwm, PWM_SYSFS_DEFAULT_PERIOD);
+
+  // init with sucess
+  return 0;
+}
+
+void pwm_sysfs_set_period(struct PWM_Sysfs *pwm, uint32_t period)
+{
+  if (write_file(pwm->period_path, "%u", period) < 0) {
+    PS_DEBUG_PRINT("failed to set period\n");
+    return;
+  }
+  PS_DEBUG_PRINT("setting period %d\n", period);
+  pwm->period_nsec = period;
+}
+
+void pwm_sysfs_set_duty(struct PWM_Sysfs *pwm, uint32_t duty)
+{
+  if (dprintf(pwm->duty_cycle_fd, "%u", duty) < 0) {
+    PS_DEBUG_PRINT("failed to set duty %u; %d; %s\n", duty, pwm->duty_cycle_fd, pwm->duty_path);
+    return;
+  }
+  pwm->duty_cycle_nsec = duty;
+  // enable pwm if not done
+  if (!pwm->enabled) {
+    pwm_sysfs_enable(pwm, true);
+  }
+}
+
+void pwm_sysfs_enable(struct PWM_Sysfs *pwm, bool enable)
+{
+  int en = enable ? 1 : 0;
+  int ret = write_file(pwm->enable_path, "%u", en);
+  if (ret < 0) {
+    PS_DEBUG_PRINT("failed to enable: %d; %s\n", ret, pwm->enable_path);
+    return;
+  }
+  PS_DEBUG_PRINT("enable channel: %d; %s\n", en, pwm->enable_path);
+  pwm->enabled = enable;
+}
+
+
+static int write_file(const char *path, const char *fmt, ...)
+{
+  errno = 0;
+
+  int fd = open(path, O_WRONLY | O_CLOEXEC);
+  if (fd == -1) {
+    return -errno;
+  }
+
+  va_list args;
+  va_start(args, fmt);
+
+  int ret = vdprintf(fd, fmt, args);
+  int errno_bkp = errno;
+  close(fd);
+
+  va_end(args);
+
+  if (ret < 1) {
+    return -errno_bkp;
+  }
+
+  return ret;
+}
+
+/*
```

Comment:
Then I would rather use a pragma to silence the warning and keep the code there...